File: leetcode/232.用栈实现队列/232.py
class MyQueue:

    def __init__(self):
        self.stackIn = []
        self.stackOut = []

    def push(self, x: int) -> None:
        self.stackIn.append(x)

    def pop(self) -> int:
        if self.empty():
            return None
        self.dumpstackIn()  # 转移数据（仅在 stackOut 为空时）
        return self.stackOut.pop()

    def peek(self) -> int:
        if self.empty():
            return None
        self.dumpstackIn()  # 转移数据（仅在 stackOut 为空时）
        return self.stackOut[-1]

    def empty(self) -> bool:
        # 队列为空当且仅当两个栈均为空
        return not self.stackIn and not self.stackOut

    def dumpstackIn(self):
        # 仅当 stackOut 为空时，从 stackIn 中转移全部数据
        if not self.stackOut:
            while self.stackIn:
                self.stackOut.append(self.stackIn.pop())

File: leetcode/232.用栈实现队列/232.java
class MyQueue {
    Stack<Integer> stackIn;
    Stack<Integer> stackOut;

    public MyQueue() {
        stackIn = new Stack<Integer>();
        stackOut = new Stack<Integer>();
    }
    
    public void push(int x) {
        stackIn.push(x);
    }
    
    public int pop() {
        dumpstackIn();
        return stackOut.pop();
    }
    
    public int peek() {
        dumpstackIn();
        return stackOut.peek();
    }
    
    public boolean empty() {
        return stackIn.isEmpty() && stackOut.isEmpty();   
    }

    public void dumpstackIn(){
        if(stackOut.isEmpty()){
            while(!stackIn.isEmpty()){
                stackOut.push(stackIn.pop());
            }
        }
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */